src/main/java/org/mule/extension/http/internal/request/HttpPollingSource.java

```diff
@@ -186,8 +216,56 @@ public void poll(PollContext<String, HttpResponseAttributes> pollContext) {
   }
 
   @Override
-  public void onRejectedItem(Result<String, HttpResponseAttributes> result, SourceCallbackContext sourceCallbackContext) {
-    LOGGER.debug("Item rejected by HTTP Polling Source in flow '{}' with result: '{}", location.getRootContainerName(),
+  public void onRejectedItem(Result<String, HttpResponseAttributes> result,
+                             SourceCallbackContext sourceCallbackContext) {
+    LOGGER.debug("Item rejected by HTTP Polling Source in flow '{}', result: '{}'", location.getRootContainerName(),
                  result.getOutput());
   }
+
+  private BindingContext buildContext(TypedValue<String> payload, HttpResponseAttributes attributes,
+                                      java.util.Optional<TypedValue<CursorStreamProvider>> item) {
+    BindingContext.Builder builder = BindingContext.builder().addBinding(PAYLOAD_PLACEGHOLDER, payload);
+    builder.addBinding(ATTRIBUTES_PLACEGHOLDER, TypedValue.of(attributes));
+    item.ifPresent(it -> builder.addBinding(ITEM_PLACEGHOLDER, it));
+    return builder.build();
+  }
+
+  private static TypedValue<String> toTypedValue(String value, MediaType mediaType, Charset encoding) {
+    if (mediaType.equals(TEXT.withCharset(mediaType.getCharset().orElse(null)))) {
+      return TypedValue.of(value);
+    } else {
+      return new TypedValue<>(value, DataType.builder().mediaType(mediaType).charset(encoding).build());
+    }
+  }
+
+  private Result<String, HttpResponseAttributes> toResult(String item, MediaType mediaType, HttpResponseAttributes attributes) {
+    return Result.<String, HttpResponseAttributes>builder().attributes(attributes).output(item).mediaType(mediaType).build();
+  }
+
+  private static boolean isJavaPayload(MediaType mediaType) {
+    return mediaType.equals(APPLICATION_JAVA.withCharset(mediaType.getCharset().orElse(null)));
+  }
+
+  private List<Result<String, HttpResponseAttributes>> getItems(InputStream fullResponse, MediaType mediaType,
+                                                                HttpResponseAttributes attributes) {
+    if (isJavaPayload(mediaType)) {
+      throw new MuleRuntimeException(createStaticMessage(format("%s is not an accepted media type",
+                                                                APPLICATION_JAVA.toRfcString())));
+    }
+    java.util.Optional<String> itemsExpression = expressions.getSplitExpression();
+    Charset charset = mediaType.getCharset().orElse(defaultCharset());
+    String response = IOUtils.toString(fullResponse, charset);
+    LOGGER.debug("Received response at {}: {} and headers {}", location.getRootContainerName(), response,
+                 attributes.getHeaders());
+    if (!itemsExpression.isPresent()) {
+      return singletonList(toResult(response, mediaType, attributes));
+    }
+    TypedValue<String> typedValue = toTypedValue(response, mediaType, charset);
+    List<Result<String, HttpResponseAttributes>> splitted = new ArrayList<>();
```

Comment:
Can you avoid creating a new list, copying all elements over, and use a stream instead? This implementation may have problems on large bodies or bodyes with many many elements.